### DISCUSSION TITLE
Path following animation with PyVista

### DISCUSSION THREAD

**Discussion by akaszynski** [pyvista/pyvista#5482] [Show and tell]
Shout out to @peterdsharpe and his excellent path following video using PyVista in [Air Racing Trajectory Optimization
](https://github.com/peterdsharpe/air-racing-optimization).

https://github.com/peterdsharpe/air-racing-optimization/assets/6295985/20478eff-b607-455e-a082-5c1ba4624a3e

**Comment by MatthewFlamm** [comment]
Super, super cool.  Thanks for sharing.  I saw in the README that one of the objectives is to reduce drastic control actions, and the impact of those control actions is immediately apparent in the videos.

**Comment by peterdsharpe** [comment]
Thanks all! :) And thanks to the PyVista team for developing such a user-friendly visualization suite!

-----

For those curious, the objective function that the optimizer minimizes is specified [here in the code](https://github.com/peterdsharpe/air-racing-optimization/blob/7874bde8bcc67011e02ddae141c4b84cf4fc2d68/threeD_continuous_optimization.py#L241):
```python
opti.minimize(
    (np.maximum(duration, 0) / 240) ** 2
    + 15 * np.mean(dyn.altitude) / 440
    + 1e-3 * wiggliness
)
```

where `wiggliness` is essentially the squared second derivative of all the various control inputs $c(t)$:

$$\mathrm{wiggliness} = \int_{t_0}^{t_f} \left(\frac{d^2 c(t)}{dt^2}\right)^2\ dt$$

-----

And the PyVista code to generate the path video [is here](https://github.com/peterdsharpe/air-racing-optimization/blob/master/final_results/final_trajectory_path_video.py).

**Reply by banesullivan** [reply]
THIS IS SO AWESOME! I've wanted this for years!
